interfaces/core.go

```diff
@@ -157,6 +157,18 @@ type Interface interface {
 	AutoConnect(plug *Plug, slot *Slot) bool
 }
 
+// Recorder describes interactions between backends and interfaces.
+type Recorder interface {
+	// RecordPermanentSlot records side-effects of having a slot.
+	RecordPermanentSlot(iface interface{}, slot *Slot) error
+	// RecordPermanentSlot records side-effects of having a plug.
+	RecordPermanentPlug(iface interface{}, plug *Plug) error
+	// RecordPermanentSlot records side-effects of having a connected slot.
+	RecordConnectedSlot(iface interface{}, plug *Plug, slot *Slot) error
+	// RecordPermanentSlot records side-effects of having a connected plug.
+	RecordConnectedPlug(iface interface{}, plug *Plug, slot *Slot) error
```

Comment:
Actually, `Add` might be better here, as it's really a request to add the given interface to the specification, rather than simply defining it.